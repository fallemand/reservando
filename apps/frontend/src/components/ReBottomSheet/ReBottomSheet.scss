@import "~@/assets/styles/variables";

.re-bottom-sheet {
  max-height: 90%;
  display: flex;
  flex-direction: column;
  padding: ($bdu * 3) 0;
  border-top-left-radius: $bdu;
  border-top-right-radius: $bdu;
  box-shadow: 0 0 $bdu/2 rgba(0, 0, 0, 0.25);

  &__title-container {
    padding: 0 ($bdu * 1.5);
    text-align: center;
    flex-grow: 1;
  }

  &__title {
    font-weight: $font-weight-medium;
    margin: 0;
  }

  &__close {
    cursor: pointer;
    color: $grape;
    text-align: left;
    margin-right: auto;
  }

  &__action {
    line-height: 1.5em;
    min-width: $bdu * 2;
    text-align: right;
  }

  &__header {
    display: flex;
    align-items: baseline;
    padding: 0 ($bdu * 2);

    &--with-actions {
      .re-bottom-sheet__title-container {
        width: 50%;
        max-width: 50%;
      }

      .re-bottom-sheet__action {
        width: 25%;
      }
    }
  }

  &__close-icon {
    display: flex;
    font-size: 14px;
    padding: $bdu;
    margin: -$bdu;
  }

  &__divider {
    margin: ($bdu * 3) (-$bdu * 2);
    border-width: 1px 0 0 0;
    border-color: $border-color;
    border-style: solid;

    &--header {
      margin-left: 0;
      margin-right: 0;
    }
  }

  &__content {
    flex-grow: 1;
    overflow-y: auto;
    padding: 0 ($bdu * 2);
  }

  &__footer {
    padding: 0 ($bdu * 2);
    margin-top: ($bdu * 3);

    &--shadow {
      padding: ($bdu * 2) ($bdu * 2) 0;
      margin-top: 0;
      z-index: $z-index-overlay;
      box-shadow: 0 (-$bdu/ 2) ($bdu / 2) rgba(0, 0, 0, 0.07);
    }
  }
}
