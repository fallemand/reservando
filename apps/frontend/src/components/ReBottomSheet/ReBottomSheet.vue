<template>
  <ReDialog
    ref="dialog"
    :class-dialog="['re-bottom-sheet', classSheet]"
    vertical-align="bottom"
    size="fullwidth"
    transition="slide-bottom"
    v-bind="$attrs"
    @close="emitClose"
    @hook:mounted="checkScrollable"
  >
    <header
      :class="[
        're-bottom-sheet__header',
        { 're-bottom-sheet__header--with-actions': $slots.action || $slots.close },
      ]"
    >
      <span class="re-bottom-sheet__close" tabindex="0" @click="close" @keydown.enter="close">
        <!-- @slot Change the default close icon -->
        <slot name="close">
          <Icon :icon="$icons.close" class="re-bottom-sheet__close-icon" />
        </slot>
      </span>

      <div v-if="title || $slots.title" class="re-bottom-sheet__title-container">
        <!-- @slot Pass a custom title. See examples -->
        <slot name="title">
          <h2 class="re-bottom-sheet__title">
            {{ title }}
          </h2>
        </slot>
      </div>

      <span v-if="$slots.action" class="re-bottom-sheet__action" tabindex="0">
        <!-- @slot Pass an action to the modal. See examples -->
        <slot name="action" />
      </span>
    </header>

    <hr class="re-bottom-sheet__divider re-bottom-sheet__divider--header" />

    <main ref="content" class="re-bottom-sheet__content">
      <!-- @slot Sheet content -->
      <slot />
    </main>

    <div
      v-if="$slots.footer"
      class="re-bottom-sheet__footer"
      :class="{ 're-bottom-sheet__footer--shadow': scrollableContent }"
    >
      <!-- @slot Sheet footer content -->
      <slot name="footer" />
    </div>
  </ReDialog>
</template>

<script lang="ts">
import Icon from "@/components/Icon/Icon.vue";
import Vue, { defineComponent } from "vue";
import ReDialog from "../ReDialog/ReDialog.vue";

const ReBottomSheet = defineComponent({
  name: "ReBottomSheet",
  components: { ReDialog, Icon },
  props: {
    /** Sheet title */
    title: {
      type: String,
      default: null,
    },
    /** CSS class to add to the sheet */
    classSheet: {
      type: [String, Array, Object],
      default: null,
    },
  },
  data() {
    return {
      scrollableContent: false,
    };
  },
  methods: {
    async checkScrollable(): Promise<void> {
      await this.$nextTick();

      const content = this.$refs.content as HTMLElement;

      if (content) {
        this.scrollableContent = content.scrollHeight > content.clientHeight;
      }
    },
    close($evt?: Event): void {
      const dialog = this.$refs.dialog as Vue & { close: ($evt?: Event) => void };
      dialog.close($evt);
    },
    emitClose($evt: Event): void {
      /** `($evt: Event)` */
      this.$emit("close", $evt);
    },
  },
});

export default ReBottomSheet;
</script>

<style lang="scss" src="./ReBottomSheet.scss" />
