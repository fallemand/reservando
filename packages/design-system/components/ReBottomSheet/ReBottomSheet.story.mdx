import { Meta, Story, ArgsTable, Canvas, Description } from "@storybook/addon-docs/blocks";
import { ReBottomSheet, ReButton } from "@reservando/design-system";

<Meta
  title="Components/BottomSheet"
  component={ReBottomSheet}
  parameters={{
    docs: {
      inlineStories: false,
    },
    viewport: {
      defaultViewport: "mobile1",
    },
  }}
/>

# ReBottomSheet

<Description of={ReBottomSheet} />

<Canvas>
  <Story name="basic" height="300px">
    {{
      components: { ReBottomSheet, ReButton },
      data() {
        return {
          show: true,
        };
      },
      template: `
      <div>
        <ReButton @click="show = true">Open sheet</ReButton>
        <ReBottomSheet v-if="show" title="Title" @close="show = false">
          <div class="example-block">
            Example content
          </div>
        </ReBottomSheet>
      </div>
    `,
    }}
  </Story>
</Canvas>

## How to use

```html
<template>
  <ReBottomSheet v-if="show" @close="show = false">Content</ReBottomSheet>
</template>

<script lang="ts">
  import { ReBottomSheet } from "@reservando/design-system";
</script>
```

<ArgsTable of={ReBottomSheet} />

It also supports all `GygDialog` props.

## Examples

#### Using the slots

<Canvas>
  <Story name="using slots" height="300px">
    {{
      components: { ReBottomSheet, ReButton },
      data() {
        return {
          show: true,
        };
      },
      methods: {
        alertAction() {
          alert("action");
        },
      },
      template: `
      <div>
        <ReButton @click="show = true">Open sheet</ReButton>
        <ReBottomSheet v-if="show" @close="show = false">
          <template #title>
            <span>Title with <b>HTML</b></span>
          </template>
          <template #action>
            <ReButton size="small" @keydown.enter="alertAction" @click="alertAction">Action</ReButton>
          </template>
          <div class="example-block">
            Example content
          </div>
        </ReBottomSheet>
      </div>
    `,
    }}
  </Story>
</Canvas>

#### Using the footer slot

The footer slot allows passing content that will be fixed to the bottom when the main content is too long. For example call to actions.
Only the main content will be scrollable.

<Canvas>
  <Story name="footer" height="350px">
    {{
      components: { ReBottomSheet, ReButton },
      data() {
        return {
          show: true,
        };
      },
      methods: {
        alertAction() {
          alert("Action");
        },
      },
      template: `
      <div>
        <ReButton @click="show = true">Open sheet</ReButton>
        <ReBottomSheet v-if="show" title="Title" @close="show=false">
          <div class="example-block" style="height: 200px;">
            Scroll the content
          </div>
          <template #footer>
            <div style="text-align: right">
              <ReButton @click="show=false" modifier="secondary" class="mr-2">Cancel</ReButton>
              <ReButton>Save</ReButton>
            </div>
          </template>
        </ReBottomSheet>
      </div>
    `,
    }}
  </Story>
</Canvas>

#### Customizing the styles

Pass `class-sheet="custom-sheet"` and use it to style your sheet

<Canvas>
  <Story name="customize styles" height="200px">
    {{
      components: { ReBottomSheet },
      data() {
        return {
          show: true,
        };
      },
      template: `
      <ReBottomSheet
        v-if="show"
        title="Title"
        class-sheet="custom-sheet"
        @close="show = false"
      >
        Content
      </ReBottomSheet>
    `,
    }}
  </Story>
</Canvas>

```css
.custom-sheet {
  height: 90%;
}
```
