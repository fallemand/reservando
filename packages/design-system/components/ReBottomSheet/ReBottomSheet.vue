<template>
  <ReDialog
    ref="dialog"
    :class-dialog="['re-bottom-sheet', classSheet]"
    vertical-align="bottom"
    size="fullwidth"
    transition="slide-bottom"
    v-bind="$attrs"
    @close="emitClose"
    @hook:mounted="checkScrollable"
  >
    <header
      :class="[
        're-bottom-sheet__header',
        { 're-bottom-sheet__header--with-actions': $slots.action || $slots.close },
      ]"
    >
      <span class="re-bottom-sheet__close" tabindex="0" @click="close" @keydown.enter="close">
        <ReIcon name="close" class="re-bottom-sheet__close-icon" size="40px" />
      </span>

      <div v-if="title || $slots.title" class="re-bottom-sheet__title-container">
        <!-- @slot Pass a custom title. See examples -->
        <slot name="title">
          <h2 class="re-bottom-sheet__title">
            {{ title }}
          </h2>
        </slot>
      </div>

      <span v-if="$slots.action" class="re-bottom-sheet__action" tabindex="0">
        <!-- @slot Pass an action to the modal. See examples -->
        <slot name="action" />
      </span>
    </header>

    <main ref="content" class="re-bottom-sheet__content">
      <!-- @slot Sheet content -->
      <slot />
    </main>

    <div
      v-if="$slots.footer"
      class="re-bottom-sheet__footer"
      :class="{ 're-bottom-sheet__footer--shadow': scrollableContent }"
    >
      <!-- @slot Sheet footer content -->
      <slot name="footer" />
    </div>
  </ReDialog>
</template>

<script lang="ts">
import Vue, { defineComponent } from "vue";
import ReIcon from "../ReIcon/ReIcon.vue";
import ReDialog from "../ReDialog/ReDialog.vue";

const ReBottomSheet = defineComponent({
  name: "ReBottomSheet",
  components: { ReDialog, ReIcon },
  props: {
    /** Sheet title */
    title: {
      type: String,
      default: null,
    },
    /** CSS class to add to the sheet */
    classSheet: {
      type: [String, Array, Object],
      default: null,
    },
  },
  data() {
    return {
      scrollableContent: false,
    };
  },
  methods: {
    async checkScrollable(): Promise<void> {
      await this.$nextTick();

      const content = this.$refs.content as HTMLElement;

      if (content) {
        this.scrollableContent = content.scrollHeight > content.clientHeight;
      }
    },
    close($evt?: Event): void {
      const dialog = this.$refs.dialog as typeof Vue & { close: ($evt?: Event) => void };
      dialog.close($evt);
    },
    emitClose($evt: Event): void {
      /** `($evt: Event)` */
      this.$emit("close", $evt);
    },
  },
});

export default ReBottomSheet;
</script>

<style lang="scss" src="./ReBottomSheet.scss" />
