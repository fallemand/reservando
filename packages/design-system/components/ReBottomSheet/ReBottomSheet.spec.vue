<script lang="ts">
import Vue from "vue";
import { shallowMount, mount } from "@vue/test-utils";
import ReBottomSheet from "./ReBottomSheet.vue";

type DialogRef = Vue & { close: ($evt?: Event) => void };

describe("ReBottomSheet component", () => {
  const props = {
    title: "__TITLE__",
    classSheet: "__CONTENT-CLASS__",
  };

  const slots = {
    default: "__MODAL-CONTENT__",
  };

  it("should render with required props", () => {
    const wrapper = shallowMount(ReBottomSheet, {
      props,
      slots,
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("using custom slots", () => {
    const wrapper = shallowMount(ReBottomSheet, {
      props,
      slots: {
        ...slots,
        action: "<button>__ACTION__</button>",
        close: "<button>__CLOSE__</button>",
        title: "<h2>__TITLE__</h2>",
        footer: "<div>__FOOTER__</div>",
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should emit close when dialog closes", async () => {
    const closeMock = jest.fn();
    const wrapper = shallowMount(ReBottomSheet, {
      props,
      methods: {},
      attrs: {
        onClose: closeMock,
      },
    });
    (wrapper.vm.$refs.dialog as DialogRef).$emit("close");
    expect(closeMock).toHaveBeenCalled();
  });

  it("should emit close when calling dialog close method", async () => {
    global.setTimeout = jest.fn();
    const closeMock = jest.fn();
    const wrapper = mount(ReBottomSheet, {
      props,
      attrs: {
        onClose: closeMock,
      },
    });
    await (wrapper.vm.$refs.dialog as DialogRef).close();
    expect(global.setTimeout).toHaveBeenCalled();
    // expect(closeMock).toHaveBeenCalled();
  });
});
</script>
