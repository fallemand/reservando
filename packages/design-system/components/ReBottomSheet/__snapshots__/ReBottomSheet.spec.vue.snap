// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReBottomSheet component should render with required props 1`] = `
<re-dialog-stub
  class-dialog="re-bottom-sheet,__CONTENT-CLASS__"
  size="fullwidth"
  name="slide-bottom"
  vertical-align="bottom"
/>
`;

exports[`ReBottomSheet component using custom slots 1`] = `
<re-dialog-stub
  class-dialog="re-bottom-sheet,__CONTENT-CLASS__"
  size="fullwidth"
  name="slide-bottom"
  vertical-align="bottom"
/>
`;
